leopard/leopard.nim

```diff
@@ -0,0 +1,263 @@
+  for i in 0..<self.workBufferCount:
+    zeroMem(self.workBufferPtr[i], self.bufSize)
+
+  # copy data into aligned buffer
+  for i in 0..<data.len:
+    copyMem(self.dataBufferPtr[i], addr data[i][0], self.bufSize)
+
+  let
+    res = leoEncode(
+      self.bufSize.cuint,
+      self.buffers.cuint,
+      self.parity.cuint,
+      self.workBufferCount.cuint,
+      cast[ptr pointer](addr self.dataBufferPtr[0]),
+      cast[ptr pointer](addr self.workBufferPtr[0]))
+
+  if ord(res) != ord(LeopardSuccess):
+    return err(leoResultString(res.LeopardResult))
+
+  for i in 0..<parity.len:
+    copyMem(addr parity[i][0], self.workBufferPtr[i], self.bufSize)
+
+  return ok()
+
+func decode*(
+  self: var LeoDecoder,
+  data,
+  parity,
+  recovered: var openArray[seq[byte]]): Result[void, cstring] =
+  ## Decode a list of buffers in `data` and `parity` into a list
+  ## of `recovered` buffers of `bufSize`. The list of `recovered`
+  ## buffers should be match the `Leo.buffers`
+  ##
+  ## `data`       - list of original data `buffers` of size `bufSize`
+  ## `parity`     - list of parity `buffers` of size `bufSize`
+  ## `recovered`  - list of recovered `buffers` of size `bufSize`
+  ##
+
+  if data.len != self.buffers:
+    return err("Number of data buffers should match!")
+
+  if parity.len != self.parity:
+    return err("Number of parity buffers should match!")
+
+  if recovered.len != self.buffers:
+    return err("Number of recovered buffers should match buffers!")
+
+  # clean out work and data buffers
+  for i in 0..<self.workBufferCount:
+    zeroMem(self.workBufferPtr[i], self.bufSize)
+
+  for i in 0..<self.decodeBufferCount:
+    zeroMem(self.decodeBufferPtr[i], self.bufSize)
+
+  for i in 0..<data.len:
+    zeroMem(self.dataBufferPtr[i], self.bufSize)
+
+  # this is needed because erasures are nil pointers
+  var
+    dataPtr = newSeq[LeoBufferPtr](data.len)
+    parityPtr = newSeq[LeoBufferPtr](self.workBufferCount)
+
+  # copy data into aligned buffer
+  for i in 0..<data.len:
+    if data[i].len > 0:
+      copyMem(self.dataBufferPtr[i], addr data[i][0], self.bufSize)
+      dataPtr[i] = self.dataBufferPtr[i]
+    else:
+      dataPtr[i] = nil
+
+  # copy parity into aligned buffer
+  for i in 0..<self.workBufferCount:
+    if i < parity.len and parity[i].len > 0:
+      copyMem(self.workBufferPtr[i], addr parity[i][0], self.bufSize)
+      parityPtr[i] = self.workBufferPtr[i]
+    else:
+      parityPtr[i] = nil
+
+  let
+    res = leo_decode(
+      self.bufSize.cuint,
+      self.buffers.cuint,
+      self.parity.cuint,
+      self.decodeBufferCount.cuint,
+      cast[ptr pointer](addr dataPtr[0]),
+      cast[ptr pointer](addr parityPtr[0]),
+      cast[ptr pointer](addr self.decodeBufferPtr[0]))
+
+  if ord(res) != ord(LeopardSuccess):
+    return err(leoResultString(res.LeopardResult))
+
+  for i, p in dataPtr:
+    if p.isNil:
+      copyMem(addr recovered[i][0], self.decodeBufferPtr[i], self.bufSize)
```

Comment:
Yes, this is how the library works, it places the recovered data in`decodeBufferPtr` located at the index where the original data was, we're then copying that into the `recovered` array or array. We can probably pass a `recovered` only large enough to contain the recovered erasures, but it's more straightforward like this, IMO.

Reply:
> Yes, this is how the library works, it places the recovered data `indecodeBufferPtr` located at the index where the original data was.

I'm aware of that aspect, e.g. see my PRs #1 (impl of higher-level `decode`) and #2 (tests). What I meant more is that instead of only copying we could...

(a) copy non-missing bytes from `data` into `recovered` so that `recovered` is the complete data set.

(b) overwrite the empty seqs in `data` with the matching indices in `recovered`, i.e. `data` is passed in missing some bytes but after successful decode the holes will be filled in.

Reply:
We could do `a`, the issue is that it's more copying for no good reason, tho it might be more ergonomic in the end. For `b`, it's a bit more complicated for two reasons:

1) we'd need an additional way of signaling where the erasures are, it could be an array of erased indexes
2) there is the question of memory ownership, for example if you look at how leopard works, it never actually allocates anything, it uses whatever the caller gives it, for a `seq` of size `0`, we'd need to allocate that memory ourself, this isn't the way the library was designed to work, nor is it good approach overall...

That said, we could use the list of erasure indexes approach and have the user pass a `data` buffer capable of holding the recovered data, in other words, without setting the data entries to `0`...

Reply:
> We could do `a`, the issue is that it's more copying for no good reason

Agreed, the additional copying seems a bit wasteful.

For `b` would this work?
```nim
# this is needed because erasures are nil pointers
var
  dataHoles = seq[int]
  dataPtr = newSeq[LeoBufferPtr](data.len)
  parityPtr = newSeq[LeoBufferPtr](self.workBufferCount)

# copy data into aligned buffer and track erasures
for i in 0..<data.len:
  if data[i].len > 0:
    copyMem(self.dataBufferPtr[i], addr data[i][0], self.bufSize)
    dataPtr[i] = self.dataBufferPtr[i]
  else:
    dataHoles.add i
    dataPtr[i] = nil
```
and at the end:
```nim
for i, p in dataPtr:
  if p.isNil:
    # copyMem(addr recovered[i][0], self.decodeBufferPtr[i], self.bufSize)
    for j in 0..<self.bufSize:
      data[i].add (self.decodeBufferPtr[i].offset j)[]
```
Will the combo of the `.add` and dereferencing result in the byte being copied so when we free the memory in `self.decodeBufferPtr` we don't have a problem?

I'm trying to answer that question for myself, but am not sure atm.

Reply:
I checked out your branch for this PR and made this adjustment:

```nim
for i, p in dataPtr:
    if p.isNil:
      copyMem(addr recovered[i][0], self.decodeBufferPtr[i], self.bufSize)
      for j in 0..<self.bufSize:
        data[i].add (cast[ptr byte](self.decodeBufferPtr[i]).offset j)[]

        debugEcho "decodeBufferPtr + offset: " & $cast[int]((cast[ptr byte](self.decodeBufferPtr[i]).offset j))
        debugEcho "addr data[i][j]: " & $cast[int](addr data[i][j])
        debugEcho "data[i][j]: " & $data[i][j]
```

I'm seeing output like this:
```
decodeBufferPtr + offset: 140473973500880
addr data[i][j]: 4498254352
data[i][j]: 19
decodeBufferPtr + offset: 140473973500881
addr data[i][j]: 4498254353
data[i][j]: 222
decodeBufferPtr + offset: 140473973500882
addr data[i][j]: 4498254354
data[i][j]: 225
```

The source and destination pointers are different, so I think we're getting an allocation + copy, but my reasoning may be off.

Reply:
Yeah, but it's not what we want because we're again trying to allocate memory inside the wrapper/library - depending on how the seq was originally created, `add` will enlarge it and potentially do a copy/move internally...

What I was thinking was more along the lines of:

```nim
func decode*(
  self: var LeoDecoder,
  data,
  parity,
  recovered: var openArray[seq[byte]],
  lostData, lostParity: seq[int]): Result[void, cstring] =
```

Then on recovery, you can call it as:

```nim
decoder.decode(data, parity, data, @[1, 2, 3], @[4, 5, 6])
```

Note the `data` being passed as `recovered`, instead of `data[i].setSize(0)`, we're simply indicating the indexes that we want recovered (same for parity)...

Reply:
We can make this change, but I want to try it out before moving forward...

Reply:
> 

Yeah, if I'm reading this correctly, thats because `add` mutates the `seq`....

Reply:
> depending on how the seq was originally created, `add` will enlarge it and potentially do a copy/move internally

When thinking about this, I had in mind L130-L135 in `leopard/leopard.nim` where the length of `data` is checked. There it's empty seqs that correspond to `dataPtr[i] = nil`. I'm having trouble understanding why calling `.add` on what we know is for sure an empty seq might result in something problematic. Even if the empty seq had been created with `newSeqOfCap` copy/move doesn't seem to be triggered by `.add` until the original cap is exceeded.

Reply:
> I'm having trouble understanding why calling .add on what we know is for sure an empty seq might result in something problematic. 

To avoid all sorts of issues with memory ownership. There is more or less an explicit contract that the api attempts to communicate to the user which says "you need to allocate and pass all the required buffers yourself", if we start breaking that promise we'll break POLA - https://en.wikipedia.org/wiki/Principle_of_least_astonishment. This a very common convention in languages without clear memory ownership rules. This is also close to how the low level leopard library operates as well and we're trying to reflect it in the higher level api as much as possible.

As for the allocations, we can't really be sure how `.add` will operate with any given combination of gc and backend, so the less we assume about the underlying behavior the safer and less fragile the library is going to be.

If you think about it the only assumption we're making is that prior to invoking encode/decode the user should have allocated enough memory for us to read/write to/from, this makes reasoning about ownership, allocations and de-allocations a lot easier than otherwise...